File: clients/java_graal/src/main/java/ai_cup_22/DebugInterface.java
package ai_cup_22;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class DebugInterface {
    private InputStream inputStream;
    private OutputStream outputStream;

    public DebugInterface(InputStream inputStream, OutputStream outputStream) {
        this.inputStream = inputStream;
        this.outputStream = outputStream;
    }

    public void addPlacedText(ai_cup_22.model.Vec2 position, String text, ai_cup_22.model.Vec2 alignment, double size, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.PlacedText(position, text, alignment, size, color));
    }
    
    public void addCircle(ai_cup_22.model.Vec2 position, double radius, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.Circle(position, radius, color));
    }
    
    public void addGradientCircle(ai_cup_22.model.Vec2 position, double radius, ai_cup_22.debugging.Color innerColor, ai_cup_22.debugging.Color outerColor) {
        this.add(new ai_cup_22.debugging.DebugData.GradientCircle(position, radius, innerColor, outerColor));
    }
    
    public void addRing(ai_cup_22.model.Vec2 position, double radius, double width, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.Ring(position, radius, width, color));
    }
    
    public void addPie(ai_cup_22.model.Vec2 position, double radius, double startAngle, double endAngle, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.Pie(position, radius, startAngle, endAngle, color));
    }
    
    public void addArc(ai_cup_22.model.Vec2 position, double radius, double width, double startAngle, double endAngle, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.Arc(position, radius, width, startAngle, endAngle, color));
    }
    
    public void addRect(ai_cup_22.model.Vec2 bottomLeft, ai_cup_22.model.Vec2 size, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.Rect(bottomLeft, size, color));
    }
    
    public void addPolygon(ai_cup_22.model.Vec2[] vertices, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.Polygon(vertices, color));
    }
    
    public void addGradientPolygon(ai_cup_22.debugging.ColoredVertex[] vertices) {
        this.add(new ai_cup_22.debugging.DebugData.GradientPolygon(vertices));
    }
    
    public void addSegment(ai_cup_22.model.Vec2 firstEnd, ai_cup_22.model.Vec2 secondEnd, double width, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.Segment(firstEnd, secondEnd, width, color));
    }
    
    public void addGradientSegment(ai_cup_22.model.Vec2 firstEnd, ai_cup_22.debugging.Color firstColor, ai_cup_22.model.Vec2 secondEnd, ai_cup_22.debugging.Color secondColor, double width) {
        this.add(new ai_cup_22.debugging.DebugData.GradientSegment(firstEnd, firstColor, secondEnd, secondColor, width));
    }
    
    public void addPolyLine(ai_cup_22.model.Vec2[] vertices, double width, ai_cup_22.debugging.Color color) {
        this.add(new ai_cup_22.debugging.DebugData.PolyLine(vertices, width, color));
    }
    
    public void addGradientPolyLine(ai_cup_22.debugging.ColoredVertex[] vertices, double width) {
        this.add(new ai_cup_22.debugging.DebugData.GradientPolyLine(vertices, width));
    }

    public void add(ai_cup_22.debugging.DebugData debugData) {
        this.send(new ai_cup_22.debugging.DebugCommand.Add(debugData));
    }
    
    public void clear() {
        this.send(new ai_cup_22.debugging.DebugCommand.Clear());
    }
    
    public void setAutoFlush(boolean enable) {
        this.send(new ai_cup_22.debugging.DebugCommand.SetAutoFlush(enable));
    }
    
    public void flush() {
        this.send(new ai_cup_22.debugging.DebugCommand.Flush());
    }

    public void send(ai_cup_22.debugging.DebugCommand command) {
        try {
            new ai_cup_22.codegame.ClientMessage.DebugMessage(command).writeTo(outputStream);
            outputStream.flush();
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public ai_cup_22.debugging.DebugState getState() {
        try {
            new ai_cup_22.codegame.ClientMessage.RequestDebugState().writeTo(outputStream);
            outputStream.flush();
            return ai_cup_22.debugging.DebugState.readFrom(inputStream);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }
}

File: clients/swift/Sources/AiCup22/DebugInterface.swift
class DebugInterface {
    private let inputStream: BufferedInputStream
    private let outputStream: BufferedOutputStream

    init(_ inputStream: BufferedInputStream, _ outputStream: BufferedOutputStream) {
        self.inputStream = inputStream
        self.outputStream = outputStream
    }

    func send(_ command: DebugCommand) {
        ClientMessage.debugMessage(command: command).writeTo(outputStream)
        outputStream.flush()
    }

    func getState() -> DebugState {
        ClientMessage.requestDebugState.writeTo(outputStream)
        outputStream.flush()
        return DebugState.readFrom(inputStream)
    }

    func addPlacedText(_ position: Vec2, _ text: String, _ alignment: Vec2, _ size: Double, _ color: Color) {
        self.add(DebugData.placedText(position: position, text: text, alignment: alignment, size: size, color: color))
    }
    func addCircle(_ position: Vec2, _ radius: Double, _ color: Color) {
        self.add(DebugData.circle(position: position, radius: radius, color: color))
    }
    func addGradientCircle(_ position: Vec2, _ radius: Double, _ innerColor: Color, _ outerColor: Color) {
        self.add(DebugData.gradientCircle(position: position, radius: radius, innerColor: innerColor, outerColor: outerColor))
    }
    func addRing(_ position: Vec2, _ radius: Double, _ width: Double, _ color: Color) {
        self.add(DebugData.ring(position: position, radius: radius, width: width, color: color))
    }
    func addPie(_ position: Vec2, _ radius: Double, _ startAngle: Double, _ endAngle: Double, _ color: Color) {
        self.add(DebugData.pie(position: position, radius: radius, startAngle: startAngle, endAngle: endAngle, color: color))
    }
    func addArc(_ position: Vec2, _ radius: Double, _ width: Double, _ startAngle: Double, _ endAngle: Double, _ color: Color) {
        self.add(DebugData.arc(position: position, radius: radius, width: width, startAngle: startAngle, endAngle: endAngle, color: color))
    }
    func addRect(_ bottomLeft: Vec2, _ size: Vec2, _ color: Color) {
        self.add(DebugData.rect(bottomLeft: bottomLeft, size: size, color: color))
    }
    func addPolygon(_ vertices: [Vec2], _ color: Color) {
        self.add(DebugData.polygon(vertices: vertices, color: color))
    }
    func addGradientPolygon(_ vertices: [ColoredVertex]) {
        self.add(DebugData.gradientPolygon(vertices: vertices))
    }
    func addSegment(_ firstEnd: Vec2, _ secondEnd: Vec2, _ width: Double, _ color: Color) {
        self.add(DebugData.segment(firstEnd: firstEnd, secondEnd: secondEnd, width: width, color: color))
    }
    func addGradientSegment(_ firstEnd: Vec2, _ firstColor: Color, _ secondEnd: Vec2, _ secondColor: Color, _ width: Double) {
        self.add(DebugData.gradientSegment(firstEnd: firstEnd, firstColor: firstColor, secondEnd: secondEnd, secondColor: secondColor, width: width))
    }
    func addPolyLine(_ vertices: [Vec2], _ width: Double, _ color: Color) {
        self.add(DebugData.polyLine(vertices: vertices, width: width, color: color))
    }
    func addGradientPolyLine(_ vertices: [ColoredVertex], _ width: Double) {
        self.add(DebugData.gradientPolyLine(vertices: vertices, width: width))
    }

    func add(_ debugData: DebugData) {
        self.send(DebugCommand.add(debugData: debugData))
    }
    func clear() {
        self.send(DebugCommand.clear)
    }
    func setAutoFlush(_ enable: Bool) {
        self.send(DebugCommand.setAutoFlush(enable: enable))
    }
    func flush() {
        self.send(DebugCommand.flush)
    }
}
